### PR TITLE
[CAD-402] Adding changes to the trace events we need and removing time from them

### PR DESCRIPTION
**Pull request by ksaric**
https://jira.iohk.io/browse/CAD-402

Let me just copy the description from there:
```
The current issue (https://jira.iohk.io/browse/CAD-314) was measuring too much.

We need to start the micro benchmarking from small and precise steps, the opposite that I started with.

A good initial microbenchmark would be the time it takes for the block to be forged during the slot the leader is producing a block.
```

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 1]
```diff
@@ -368,11 +377,10 @@ forkBlockProduction maxBlockBodySize IS{..} BlockProduction{..} =
...
           curTip <- atomically $ ChainDB.getTipPoint chainDB
           if curTip == blockPoint newBlock then do
-            -- We measure the time directly.
-            -- We could also measure it when collecting the trace events, but this
-            -- is much more precise.
-            time <- getMonotonicTime
-            trace $ TraceAdoptedBlock currentSlot newBlock txs time
+            -- There is no need to measure the time directly, the precision of
+            -- adding the actual time measurement on the contravariant tracer is
+            -- the same.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: I appreciate the comment, but outside the context of this PR, it seems kind of strange to talk about something that we *used* to do. I think a comment in the PR itself would have made more sense. So I'd remove it.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 2]
```diff
@@ -328,6 +329,14 @@ forkBlockProduction maxBlockBodySize IS{..} BlockProduction{..} =
             case mIsLeader of
               Nothing    -> return NotLeader
               Just proof -> do
+                -- We need to measure the trace just before we actually run the
+                -- protocol because it is possible that the the __current node
+                -- is not the slot leader__ and will __not lead__ the protocol/block
+                -- production.
+                --
+                -- TODO(KS): We need to add the mempool size for measurement.
+                pure . trace $ TraceForgeAboutToLead currentSlot
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Hmm, we're inside a `STM` here, so you can't trace here. This `pure` means you ignore the action to trace it :confused:

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 3]
```diff
@@ -328,6 +329,14 @@ forkBlockProduction maxBlockBodySize IS{..} BlockProduction{..} =
             case mIsLeader of
               Nothing    -> return NotLeader
               Just proof -> do
+                -- We need to measure the trace just before we actually run the
+                -- protocol because it is possible that the the __current node
+                -- is not the slot leader__ and will __not lead__ the protocol/block
+                -- production.
+                --
+                -- TODO(KS): We need to add the mempool size for measurement.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: No need for this TODO, we have a ticket open for it (#1431) that will shortly be addressed.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Also, now you're tracing `TraceForgeAboutToLead` twice.

Comment (ksaric) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Yep, forgot to remove that one :grin:

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 4]
```diff
@@ -311,7 +311,14 @@ forkBlockProduction maxBlockBodySize IS{..} BlockProduction{..} =
...
           withTxs = withSyncState mempool (TxsForBlockInSlot currentSlot)
 
+      -- We need to measure the trace just before we actually run the
+      -- protocol because it is possible that the the __current node
+      -- is not the slot leader__ and will __not lead__ the protocol/block
+      -- production.
+      -- Currently, we will simply drop the unmatched trace event and take
+      -- the diff between the last event of this type and the forged block event.
       trace $ TraceForgeAboutToLead currentSlot
+
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: In fact, do we need this comment here? This seems like something that belongs in the monitoring/benchmarking code?

Comment (ksaric) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Removed.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 5]
```diff
@@ -312,6 +312,7 @@ forkBlockProduction maxBlockBodySize IS{..} BlockProduction{..} =
           withTxs = withSyncState mempool (TxsForBlockInSlot currentSlot)
 
       trace $ TraceForgeAboutToLead currentSlot
+
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Nitpicking: can you remove this newly added blank line?

**Pull request merged**